presto-common/src/main/java/com/facebook/presto/common/block/AbstractMapBlock.java

```diff
@@ -0,0 +1,517 @@
+        }
+        return size;
+    }
+
+    @Override
+    public boolean mayHaveNull()
+    {
+        return getMapIsNull() != null;
+    }
+
+    @Override
+    public boolean isNull(int position)
+    {
+        checkReadablePosition(position);
+        boolean[] mapIsNull = getMapIsNull();
+        return mapIsNull != null && mapIsNull[position + getOffsetBase()];
+    }
+
+    public boolean isHashTablesPresent()
+    {
+        return getHashTables().get() != null;
+    }
+
+    private void checkReadablePosition(int position)
+    {
+        if (position < 0 || position >= getPositionCount()) {
+            throw new IllegalArgumentException("position is not valid");
+        }
+    }
+
+    public static class HashTables
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(HashTables.class).instanceSize();
+
+        // Hash to location in map. Writes to the field by MapBlock is protected by "HashTables" monitor in MapBlock.
+        // MapBlockBuilder instances have their dedicated hashTables instances, so the write accesses to the hashTables
+        // fields do not need to be synchronized in that class.
+        @Nullable
+        private volatile int[] hashTables;
+
+        // The number of hash tables. Each map row corresponds to one hash table if it's built.
+        private int expectedHashTableCount;
+
+        HashTables(Optional<int[]> hashTables, int expectedHashTableCount)
+        {
+            this.hashTables = hashTables.orElse(null);
+            this.expectedHashTableCount = expectedHashTableCount;
+        }
+
+        @Nullable
+        int[] get()
+        {
+            return hashTables;
+        }
+
+        void set(int[] hashTables)
+        {
+            requireNonNull(hashTables, "hashTables is null");
+            this.hashTables = hashTables;
+        }
+
+        void setExpectedHashTableCount(int count)
+        {
+            expectedHashTableCount = count;
+        }
+
+        int getExpectedHashTableCount()
+        {
+            return expectedHashTableCount;
+        }
+
+        public long getRetainedSizeInBytes()
+        {
+            return INSTANCE_SIZE + sizeOf(hashTables);
+        }
+
+        public void loadHashTables(int positionCount, int[] offsets, boolean[] mapIsNull, Block keyBlock, MethodHandle keyBlockHashCode)
+        {
+            int[] hashTables = new int[keyBlock.getPositionCount() * HASH_MULTIPLIER];
+            Arrays.fill(hashTables, -1);
+
+            verify(positionCount < offsets.length, "incorrect offsets size");
+
+            for (int i = 0; i < positionCount; i++) {
+                int keyOffset = offsets[i];
+                int keyCount = offsets[i + 1] - keyOffset;
+                if (keyCount < 0) {
+                    throw new IllegalArgumentException(format("Offset is not monotonically ascending. offsets[%s]=%s, offsets[%s]=%s", i, offsets[i], i + 1, offsets[i + 1]));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: &#39;arrayget()&#39; expression, &#39;+&#39; expression, i (declared at line 451). This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 455. Consider formatting this data with &#39;%d&#39; instead.